src/engine/waypoint_system.cpp

```diff
@@ -97,8 +77,159 @@ void WaypointSystem::Update(double DeltaTime)
 }
 
 
+void WaypointSystem::OnWaypointSelect( std::string waypointName )
+{
+    EventServer<WaypointChangedEvent>::Get().SendEvent({ 
+        waypointName, 
+        waypointName == NextStage ? mWaypointsData.mPlayerToWaypoint[mProgramState.mControlledActorGUID] : mWaypointsData.mWaypointGUIDs[waypointName],
+        mProgramState.mControlledActorGUID,
+        WaypointChangedEvent::Selected } );
+}
+
 void WaypointSystem::OnMapStart( core::MapStartEvent const& Evt )
 {
+    if (Evt.mState == core::MapStartEvent::Started)
+    {
+        mWaypointsData.mNextWaypointCounter = 1;
+        mWaypointsData.mWaypointNames = { NextStage, Resume };
+        mWaypointsData.mWaypointGUIDs = { { NextStage, -1 }, { Resume,-1 } };
+        mWaypointsData.mPlayerToWaypoint.clear();
+    }
+}
+
+
+void WaypointSystem::OnWaypointChanged( WaypointChangedEvent const& Evt )
+{
+    if (Evt.mState == WaypointChangedEvent::Lit)
+    {
+        mWaypointsData.mWaypointNames.push_back( Evt.mWaypointName );
+        mWaypointsData.mWaypointGUIDs[Evt.mWaypointName] = Evt.mWaypointGUID;
+        ++mWaypointsData.mNextWaypointCounter;
+        auto waypoint( mScene.GetActor( Evt.mWaypointGUID ) );
+        if (waypoint.IsValid())
+        {
+            auto waypointC( waypoint->Get<IWaypointComponent>());
+            if (waypointC.IsValid())
+            {
+                waypointC->SetLit( true );
+            }
+        }
+    }
+    else if (Evt.mState == WaypointChangedEvent::Choose)
+    {
+        auto player( mScene.GetActor( Evt.mPlayerGUID ) );
+        if (player.IsValid())
+        {
+            Opt<PlayerControllerComponent> pcc( player->Get<IControllerComponent>() );
+            if (pcc.IsValid())
+            {
+                pcc->mActivate.Handled();
+                pcc->SetEnabled( false );
```

Comment:
I'd prefer to use a similar solution to the machine gun root mechanism. Players will get stuck. Players don't like being stuck.